packages/core/src/lib/utils/webauthn-utils.ts

```diff
@@ -0,0 +1,418 @@
+import type { MaybeUserWithID, WebAuthnProviderType } from "../../providers/webauthn";
+import type { Account, Authenticator, Awaited, InternalOptions, RequestInternal, ResponseInternal } from "../../types";
+import type { Cookie } from "./cookie";
+import { AdapterError, AuthError, InvalidProvider, MissingAdapter, WebAuthnVerificationError } from "../../errors";
+import { webauthnChallenge } from "../actions/callback/oauth/checks";
+import { VerifiedAuthenticationResponse, VerifiedRegistrationResponse, generateAuthenticationOptions, generateRegistrationOptions, verifyAuthenticationResponse, verifyRegistrationResponse } from "@simplewebauthn/server";
+import type {
+  AuthenticationResponseJSON,
+  PublicKeyCredentialCreationOptionsJSON,
+  PublicKeyCredentialRequestOptionsJSON,
+  RegistrationResponseJSON,
+} from "@simplewebauthn/server/script/deps"
+import type { Adapter, AdapterAccount, AdapterAuthenticator, AdapterUser } from "../../adapters";
+import type { GetUserInfo } from "../../providers/webauthn";
+import { randomString } from "./web";
+
+export type WebAuthnRegister = "register"
+export type WebAuthnAuthenticate = "authenticate"
+export type WebAuthnAction = WebAuthnRegister | WebAuthnAuthenticate
+
+type SessionUser = Pick<AdapterUser, "name" | "email" | "id">
+
+type InternalOptionsWebAuthn = InternalOptions<WebAuthnProviderType> & { adapter: Required<Adapter> }
+export type WebAuthnOptionsResponseBody = {
+  action: WebAuthnAuthenticate,
+  options: PublicKeyCredentialRequestOptionsJSON
+} | {
+  action: WebAuthnRegister,
+  options: PublicKeyCredentialCreationOptionsJSON
+}
+type WebAuthnOptionsResponse = ResponseInternal & {
+  body: WebAuthnOptionsResponseBody
+}
+
+export type CredentialDeviceType = "singleDevice" | "multiDevice"
+interface InternalAuthenticator {
+  providerAccountId: string
+  credentialID: Uint8Array
+  credentialPublicKey: Uint8Array
+  counter: number
+  credentialDeviceType: CredentialDeviceType
+  credentialBackedUp: boolean
+  transports?: AuthenticatorTransport[]
+}
+
+type RGetUserInfo = Awaited<ReturnType<GetUserInfo>>
+
+/**
+ * Decides the WebAuthn options based on the provided parameters.
+ * 
+ * @param action - The WebAuthn action to perform (optional).
+ * @param loggedInUser - The logged-in user (optional).
+ * @param userInfoResponse - The response containing user information (optional).
+ * 
+ * @returns The WebAuthn action to perform, or undefined if no decision could be made.
```

Comment:
shouldn't it be null
```suggestion
 * @returns The WebAuthn action to perform, or null if no decision could be made.
```